src/common/components/hosts/MassChangeHostnameModal.tsx

```diff
@@ -0,0 +1,280 @@
+  const inventory = stringToJSON<Inventory>(inventoryString) || {};
+  return getHostnameUtils(host, inventory) || '';
+};
+
+const getHostnameTemplateAndCount = (values: EditHostFormValues) => {
+  const numberTemplate = values.hostname.match(/{{n+}}/) || [];
+  const numberCount = numberTemplate[0]?.match(/n/g)?.length || 0;
+  return { numberTemplate, numberCount };
+};
+
+const templateToHostname = (
+  index: number,
+  values: EditHostFormValues,
+  numberTemplate: RegExpMatchArray,
+  numberCount: number,
+) => values.hostname.replace(numberTemplate[0], `${index + 1}`.padStart(numberCount, '0'));
+
+const getNewHostnames = (values: EditHostFormValues, selectedHosts: Host[]) => {
+  const { numberTemplate, numberCount } = getHostnameTemplateAndCount(values);
+  return selectedHosts.map((h, index) =>
+    templateToHostname(index, values, numberTemplate, numberCount),
+  );
+};
+
+type EditHostFormValues = {
+  hostname: string;
+};
+
+const initialValues = {
+  hostname: '',
+};
+
+const validationSchema = (initialValues: EditHostFormValues, usedHostnames: string[]) =>
+  Yup.object().shape({
+    hostname: hostnameValidationSchema(initialValues.hostname, usedHostnames),
+  });
+
+const withTemplate = (
+  selectedHosts: Host[],
+  hosts: Host[],
+  schema: ReturnType<typeof validationSchema>,
+) => async (values: EditHostFormValues) => {
+  const newHostnames = getNewHostnames(values, selectedHosts);
+
+  const usedHostnames = hosts.reduce<string[]>((acc, host) => {
+    if (!selectedHosts.find((a) => a.id === host.id)) {
+      acc.push(getHostname(host));
+    }
+    return acc;
+  }, []);
+  let validationResult = await getRichTextValidation(schema)({
+    ...values,
+    hostname: newHostnames[0],
+  });
+  if (newHostnames.some((h) => usedHostnames.includes(h))) {
+    validationResult = {
+      ...(validationResult || {}),
+      hostname: (validationResult?.hostname || []).concat(HOSTNAME_VALIDATION_MESSAGES.NOT_UNIQUE),
+    };
+  }
+  return validationResult;
+};
+
+type MassChangeHostnameFormProps = {
+  selectedHosts: Host[];
+  isOpen: boolean;
+  onClose: VoidFunction;
+  patchingHost: number;
+};
+
+const MassChangeHostnameForm: React.FC<MassChangeHostnameFormProps> = ({
+  selectedHosts: initHosts,
+  isOpen,
+  patchingHost,
+  onClose,
+}) => {
+  const { values, handleSubmit, isSubmitting, status, isValid } = useFormikContext<
+    EditHostFormValues
+  >();
+
+  const hostnameInputRef = React.useRef<HTMLInputElement>();
+  const ref = React.useRef<Host[]>(initHosts);
+
+  React.useEffect(() => {
+    isOpen && hostnameInputRef.current?.focus();
+  }, [isOpen]);
+
+  React.useCallback(() => {
+    if (!isSubmitting) {
+      ref.current = initHosts;
+    }
+  }, [initHosts, isSubmitting]);
+
+  const selectedHosts = ref.current;
+
+  const newHostnames = getNewHostnames(values, selectedHosts);
+
+  return (
+    <Form onSubmit={handleSubmit}>
+      <div>
+        <ModalBoxBody>
```

Comment:
I think this dialog can be generalized and reused for each case.

Reply:
TBH Im not sure what to generalize here.
Modal body is unique for every modal
Formik props are unique too.

So..the only thing missing is showing progress/alert - i created a generic component `ModalProgress` for that. If there is anything else, let me know

Reply:
I was thinking about the modal dialog structure, it's repeating.
But when looking closer, it would be just a little gain not outweighing having a new abstraction layer. The feature-specific code is significantly bigger than that potentially shared part.  So let's keep it as it is.